pkg/terraform/terraform.go

```diff
@@ -335,19 +338,68 @@ func (t *Terraform) command(cluster interfaces.Cluster, args []string, stdin io.
 	return err
 }
 
-func (t *Terraform) Plan(cluster interfaces.Cluster) error {
-	return t.terraformWrapper(
+func (t *Terraform) terraformPlanPath(cluster interfaces.Cluster) string {
+	return filepath.Join(t.codePath(cluster), terraformPlanFile)
+}
+
+// this checks if an error is coming from an exec failing with exit code 2,
+// this is typicall for a terraform plan that has changes
+func errIsTerraformPlanChangesNeeded(err error) bool {
+	if err == nil {
+		return false
+	} else if exitError, ok := err.(*exec.ExitError); !ok {
+		return false
+	} else if status, ok := exitError.ProcessState.Sys().(syscall.WaitStatus); !ok {
+		return false
+	} else if status.ExitStatus() != 2 {
+		return false
+	}
+
+	return true
+}
+
+func (t *Terraform) Plan(cluster interfaces.Cluster) (changesNeeded bool, err error) {
+	err = t.terraformWrapper(
 		cluster,
 		"plan",
-		[]string{"-detailed-exitcode", "-input=false"},
+		[]string{"-detailed-exitcode", "-input=false", fmt.Sprintf("-out=%s", terraformPlanFile)},
 	)
+	changesNeeded = errIsTerraformPlanChangesNeeded(err)
+	if err != nil && !changesNeeded {
+		return false, err
+	}
+
+	tfPlan, err := plan.Open(t.terraformPlanPath(cluster))
+	if err != nil {
+		return false, fmt.Errorf("error while trying to read plan file: %s", err)
+	}
+
+	isDestroyingEBSVolume, ebsVolumesToDestroy := plan.IsDestroyingEBSVolume(tfPlan)
+	t.log.Warnf("plan output %+v %+v", isDestroyingEBSVolume, ebsVolumesToDestroy)
+	if isDestroyingEBSVolume {
+		// TODO: add override flag
+		return false, fmt.Errorf("error this will destroy the following EBS volumes: %s", strings.Join(ebsVolumesToDestroy, ", "))
+	}
+
+	return changesNeeded, nil
 }
 
 func (t *Terraform) Apply(cluster interfaces.Cluster) error {
+	// TODO: handle supplied plan
+
+	// generate a plan
+	if changesNeeded, err := t.Plan(cluster); err != nil {
+		return err
+	} else if !changesNeeded {
+		// nothing to do
+		return nil
+	}
+
```

Comment:
A check if ctx if done here could be nice

Reply:
Looking at that now